// flairup/dist/index.js
"use strict";
function _array_like_to_array(arr, len) {
    if (len == null || len > arr.length) len = arr.length;
    for(var i = 0, arr2 = new Array(len); i < len; i++)arr2[i] = arr[i];
    return arr2;
}
function _array_with_holes(arr) {
    if (Array.isArray(arr)) return arr;
}
function _array_without_holes(arr) {
    if (Array.isArray(arr)) return _array_like_to_array(arr);
}
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _create_class(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
function _instanceof(left, right) {
    if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) {
        return !!right[Symbol.hasInstance](left);
    } else {
        return left instanceof right;
    }
}
function _iterable_to_array(iter) {
    if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter);
}
function _iterable_to_array_limit(arr, i) {
    var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"];
    if (_i == null) return;
    var _arr = [];
    var _n = true;
    var _d = false;
    var _s, _e;
    try {
        for(_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true){
            _arr.push(_s.value);
            if (i && _arr.length === i) break;
        }
    } catch (err) {
        _d = true;
        _e = err;
    } finally{
        try {
            if (!_n && _i["return"] != null) _i["return"]();
        } finally{
            if (_d) throw _e;
        }
    }
    return _arr;
}
function _non_iterable_rest() {
    throw new TypeError("Invalid attempt to destructure non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _non_iterable_spread() {
    throw new TypeError("Invalid attempt to spread non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _sliced_to_array(arr, i) {
    return _array_with_holes(arr) || _iterable_to_array_limit(arr, i) || _unsupported_iterable_to_array(arr, i) || _non_iterable_rest();
}
function _to_consumable_array(arr) {
    return _array_without_holes(arr) || _iterable_to_array(arr) || _unsupported_iterable_to_array(arr) || _non_iterable_spread();
}
function _unsupported_iterable_to_array(o, minLen) {
    if (!o) return;
    if (typeof o === "string") return _array_like_to_array(o, minLen);
    var n = Object.prototype.toString.call(o).slice(8, -1);
    if (n === "Object" && o.constructor) n = o.constructor.name;
    if (n === "Map" || n === "Set") return Array.from(n);
    if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _array_like_to_array(o, minLen);
}
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = function(target, all) {
    for(var name in all)__defProp(target, name, {
        get: all[name],
        enumerable: true
    });
};
var __copyProps = function(to, from, except, desc) {
    if (from && typeof from === "object" || typeof from === "function") {
        var _iteratorNormalCompletion = true, _didIteratorError = false, _iteratorError = undefined;
        try {
            var _loop = function() {
                var key = _step.value;
                if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
                    get: function() {
                        return from[key];
                    },
                    enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
                });
            };
            for(var _iterator = __getOwnPropNames(from)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true)_loop();
        } catch (err) {
            _didIteratorError = true;
            _iteratorError = err;
        } finally{
            try {
                if (!_iteratorNormalCompletion && _iterator.return != null) {
                    _iterator.return();
                }
            } finally{
                if (_didIteratorError) {
                    throw _iteratorError;
                }
            }
        }
    }
    return to;
};
var __toCommonJS = function(mod) {
    return __copyProps(__defProp({}, "__esModule", {
        value: true
    }), mod);
};
// src/index.ts
var src_exports = {};
__export(src_exports, {
    createSheet: function() {
        return createSheet;
    },
    cx: function() {
        return cx;
    }
});
module.exports = __toCommonJS(src_exports);
// src/utils/is.ts
var is = {
    pseudoSelector: function(selector, _) {
        return selector.startsWith(":");
    },
    mediaQuery: function(property, _) {
        return property.startsWith("@media");
    },
    directClass: function(property, _) {
        return property === ".";
    },
    cssVariables: function(property, _) {
        return property === "--";
    },
    validProperty: function(property, value) {
        return (typeof value === "string" || typeof value === "number") && !is.cssVariables(property, value) && !is.pseudoSelector(property, value) && !is.mediaQuery(property, value);
    },
    topLevelClass: function(property, _) {
        return property.startsWith(".") && property.length > 1;
    },
    string: function(value) {
        return typeof value === "string";
    }
};
// src/utils/stableHash.ts
function stableHash(prefix, seed) {
    var hash = 0;
    if (seed.length === 0) return hash.toString();
    for(var i = 0; i < seed.length; i++){
        var char = seed.charCodeAt(i);
        hash = (hash << 5) - hash + char;
        hash = hash & hash;
    }
    return "".concat(prefix !== null && prefix !== void 0 ? prefix : "cl", "_").concat(hash.toString(36));
}
// src/utils/stringManipulators.ts
function genCssRule(classes, property, value) {
    return genCssRules(classes, genLine(property, value));
}
function genLine(property, value) {
    return "".concat(camelCaseToDash(property), ": ").concat(handlePropertyValue(property, value), ";");
}
function handlePropertyValue(property, value) {
    if (property === "content") {
        return '"'.concat(value, '"');
    }
    return value;
}
function genCssRules(classes, content) {
    return "".concat(makeClassName(classes), " ").concat(wrapWithCurlys(content));
}
function wrapWithCurlys(content) {
    var breakLine = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : false;
    return [
        breakLine ? "{\n" : "{",
        content,
        breakLine ? "\n}" : "}"
    ].join("");
}
function camelCaseToDash(str) {
    return str.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
}
function joinedProperty(property, value) {
    return "".concat(property, ":").concat(value);
}
function chunkSelector(className, property) {
    var base = makeClassName(className);
    if (is.pseudoSelector(property)) {
        return "".concat(base).concat(property);
    }
    if (is.mediaQuery(property)) {
        return "".concat(property);
    }
    return base;
}
function makeClassName(classes) {
    return classes.filter(Boolean).map(function(c) {
        return ".".concat(c);
    }).join(" ");
}
function appendString(base, line) {
    return base ? "".concat(base, "\n").concat(line) : line;
}
// src/Sheet.ts
var Sheet = /*#__PURE__*/ function() {
    function Sheet(name) {
        _class_call_check(this, Sheet);
        this.name = name;
        // Hash->css
        this.storedStyles = {};
        // styles->hash
        this.storedClasses = {};
        this.style = "";
        this.count = 0;
        this.id = "flairup-".concat(name);
        this.styleTag = this.createStyleTag();
    }
    _create_class(Sheet, [
        {
            key: "getStyle",
            value: function getStyle() {
                return this.style;
            }
        },
        {
            key: "append",
            value: function append(css) {
                this.style = appendString(this.style, css);
            }
        },
        {
            key: "apply",
            value: function apply() {
                this.count++;
                if (!this.styleTag) {
                    return;
                }
                this.styleTag.innerHTML = this.style;
            }
        },
        {
            key: "isApplied",
            value: function isApplied() {
                return !!this.styleTag;
            }
        },
        {
            key: "createStyleTag",
            value: function createStyleTag() {
                if (typeof document === "undefined" || this.isApplied()) {
                    return;
                }
                var styleTag = document.createElement("style");
                styleTag.type = "text/css";
                styleTag.id = this.id;
                document.head.appendChild(styleTag);
                return styleTag;
            }
        },
        {
            key: "addRule",
            value: function addRule(property, value) {
                var parentClassName = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : "";
                var key = joinedProperty(property, value) + parentClassName;
                var storedClass = this.storedClasses[key];
                if (is.string(storedClass)) {
                    return storedClass;
                }
                var hash = stableHash(this.name, key);
                this.storedClasses[key] = hash;
                this.storedStyles[hash] = [
                    property,
                    value
                ];
                this.append(genCssRule([
                    parentClassName,
                    hash
                ], property, value));
                return hash;
            }
        }
    ]);
    return Sheet;
}();
// src/utils/forIn.ts
function forIn(obj, fn) {
    for(var key in obj){
        fn(key.trim(), obj[key]);
    }
}
// src/cx.ts
function cx() {
    for(var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++){
        args[_key] = arguments[_key];
    }
    var classes = args.reduce(function(classes2, arg) {
        if (_instanceof(arg, Set)) {
            var _classes2;
            (_classes2 = classes2).push.apply(_classes2, _to_consumable_array(arg));
        } else if (typeof arg === "string") {
            classes2.push(arg);
        } else if (Array.isArray(arg)) {
            classes2.push(cx.apply(void 0, _to_consumable_array(arg)));
        } else if (typeof arg === "object") {
            Object.entries(arg).forEach(function(param) {
                var _param = _sliced_to_array(param, 2), key = _param[0], value = _param[1];
                if (value) {
                    classes2.push(key);
                }
            });
        }
        return classes2;
    }, []);
    return classes.filter(Boolean).join(" ").trim();
}
// src/index.ts
function createSheet(name) {
    var sheet = new Sheet(name);
    return {
        create: create,
        getStyle: sheet.getStyle.bind(sheet),
        isApplied: sheet.isApplied.bind(sheet)
    };
    function create(styles) {
        var scopedStyles = {};
        forIn(styles, function(scopeName, styles2) {
            var scopeClassName = stableHash(sheet.name, scopeName);
            if (is.topLevelClass(scopeName, styles2)) {
                forIn(styles2, function(property, value) {
                    iterateStyles(sheet, value, scopeClassName, scopeName.slice(1)).forEach(function(className) {
                        addScopedStyle(property, className);
                    });
                });
                return;
            }
            iterateStyles(sheet, styles2, scopeClassName).forEach(function(className) {
                addScopedStyle(scopeName, className);
            });
        });
        sheet.apply();
        return scopedStyles;
        function addScopedStyle(name2, className) {
            var _scopedStyles_name2;
            scopedStyles[name2] = (_scopedStyles_name2 = scopedStyles[name2]) !== null && _scopedStyles_name2 !== void 0 ? _scopedStyles_name2 : /* @__PURE__ */ new Set();
            scopedStyles[name2].add(className);
        }
    }
}
function iterateStyles(sheet, styles, scopeClassName, parentClassName) {
    var output = /* @__PURE__ */ new Set();
    forIn(styles, function(property, value) {
        if (is.directClass(property, value)) {
            return handleAddedClassnames(value).forEach(function(classes) {
                return output.add(classes);
            });
        }
        if (is.mediaQuery(property, value)) {
            return handleMediaQuery(sheet, value, property, scopeClassName).forEach(function(className) {
                return output.add(className);
            });
        }
        if (is.pseudoSelector(property, value) || is.cssVariables(property, value)) {
            return handleChunks(sheet, value, property, scopeClassName).forEach(function(classes) {
                return output.add(classes);
            });
        }
        if (is.validProperty(property, value)) {
            var ruleClassName = sheet.addRule(property, value, parentClassName);
            return output.add(ruleClassName);
        }
    });
    return output;
}
function handleAddedClassnames(classes) {
    return [].concat(classes);
}
function handleChunks(sheet, styles, property, scopeClassName) {
    var classes = /* @__PURE__ */ new Set();
    var chunkRows = [];
    forIn(styles, function(property2, value) {
        if (is.validProperty(property2, value)) {
            chunkRows.push(genLine(property2, value));
            return;
        }
        iterateStyles(sheet, value !== null && value !== void 0 ? value : {}, scopeClassName).forEach(function(className) {
            return classes.add(className);
        });
    });
    if (chunkRows.length) {
        var output = chunkRows.join(" ");
        sheet.append("".concat(chunkSelector([
            scopeClassName
        ], property), " ").concat(wrapWithCurlys(output, true)));
    }
    classes.add(scopeClassName);
    return classes;
}
function handleMediaQuery(sheet, styles, property, scopeClassName) {
    sheet.append(chunkSelector([
        scopeClassName
    ], property) + " {");
    var output = iterateStyles(sheet, styles, scopeClassName);
    sheet.append("}");
    return output;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
    createSheet: createSheet,
    cx: cx
});
